chirp/chirp.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package chirp implements a client to communicate with BIRD.
+package chirp
+
+import (
+	"fmt"
+	"net"
+	"strings"
+)
+
+// New creates a BIRDClient.
+func New(socket string) (*BIRDClient, error) {
+	conn, err := newConn(socket)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to BIRD: %w", err)
+	}
+	return &BIRDClient{socket: socket, conn: conn}, nil
+}
+
+// BIRDClient handles communication with BIRD.
+type BIRDClient struct {
+	socket string
+	conn   net.Conn
+}
+
+// DisableProtocol disables the provided protocol.
+func (b *BIRDClient) DisableProtocol(protocol string) error {
+	out, err := b.exec(fmt.Sprintf("disable %s\n", protocol))
+	if err != nil {
+		return err
+	}
+	if strings.Contains(out, fmt.Sprintf("%s: already disabled", protocol)) {
+		return nil
+	} else if strings.Contains(out, fmt.Sprintf("%s: disabled", protocol)) {
+		return nil
+	}
+	return fmt.Errorf("failed to disable %s: %v", protocol, out)
+}
+
+// EnableProtocol enables the provided protocol.
+func (b *BIRDClient) EnableProtocol(protocol string) error {
+	out, err := b.exec(fmt.Sprintf("enable %s\n", protocol))
+	if err != nil {
+		return err
+	}
+	if strings.Contains(out, fmt.Sprintf("%s: already enabled", protocol)) {
+		return nil
+	} else if strings.Contains(out, fmt.Sprintf("%s: enabled", protocol)) {
+		return nil
+	}
+	return fmt.Errorf("failed to enable %s: %v", protocol, out)
+}
+
+func (b *BIRDClient) exec(cmd string) (string, error) {
+	if _, err := b.conn.Write([]byte(cmd)); err != nil {
+		return "", err
+	}
+	buf := make([]byte, 4096)
+	if _, err := b.conn.Read(buf); err != nil {
```

Comment:
this isn't a safe way to read from a net.Conn. there are zero guarantees about how much will be returned. you need to loop.

I assume the output is framed somehow? with a length? with a newline character?

Reply:
Would a `bufio.Scanner` work?